processing/src/main/java/io/druid/query/search/SearchQueryRunner.java

```diff
@@ -79,44 +85,61 @@ public SearchQueryRunner(Segment segment)
 
     final SearchQuery query = (SearchQuery) input;
     final Filter filter = Filters.convertDimensionFilters(query.getDimensionsFilter());
-    final List<String> dimensions = query.getDimensions();
+    final List<DimensionSpec> dimensions = query.getDimensions();
     final SearchQuerySpec searchQuerySpec = query.getQuery();
     final int limit = query.getLimit();
 
+    // Closing this will cause segfaults in unit tests.
     final QueryableIndex index = segment.asQueryableIndex();
+
     if (index != null) {
       final TreeSet<SearchHit> retVal = Sets.newTreeSet(query.getSort().getComparator());
 
-      Iterable<String> dimsToSearch;
+      Iterable<DimensionSpec> dimsToSearch;
       if (dimensions == null || dimensions.isEmpty()) {
-        dimsToSearch = index.getAvailableDimensions();
+        dimsToSearch = Iterables.transform(index.getAvailableDimensions(), Druids.DIMENSION_IDENTITY);
       } else {
         dimsToSearch = dimensions;
       }
 
-      BitmapFactory bitmapFactory = index.getBitmapFactoryForDimensions();
+      final BitmapFactory bitmapFactory = index.getBitmapFactoryForDimensions();
 
       final ImmutableBitmap baseFilter;
       if (filter == null) {
         baseFilter = bitmapFactory.complement(bitmapFactory.makeEmptyImmutableBitmap(), index.getNumRows());
       } else {
-        ColumnSelectorBitmapIndexSelector selector = new ColumnSelectorBitmapIndexSelector(bitmapFactory, index);
+        final ColumnSelectorBitmapIndexSelector selector = new ColumnSelectorBitmapIndexSelector(bitmapFactory, index);
         baseFilter = filter.getBitmapIndex(selector);
       }
 
-      for (String dimension : dimsToSearch) {
-        final Column column = index.getColumn(dimension);
+      for (DimensionSpec dimension : dimsToSearch) {
+        final Column column = index.getColumn(dimension.getDimension());
         if (column == null) {
           continue;
         }
 
         final BitmapIndex bitmapIndex = column.getBitmapIndex();
+        final Function<String, String> extrFn;
+        final ExtractionFn extractionFn = dimension.getExtractionFn();
+        if (extractionFn != null) {
+          extrFn = new Function<String, String>()
```

Comment:
it's confusing to have both extrFn and extractionFn, is it necessary to wrap the extraction function in another function? Maybe we should just define an identity extraction function to make this easier.